llvm/lib/Transforms/Utils/LoopIterCountAssumptions.cpp

```diff
@@ -0,0 +1,224 @@
+//===-- LoopIterCountAssumptions.cpp - add Loop assumptions -----*- C++ -*-===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// (c) Copyright 2024 Advanced Micro Devices, Inc. or its affiliates
+//
+//===----------------------------------------------------------------------===//
+//
+// This pass converts Loop Iteration Count Metadata to Assumptions which can be
+// picked up by Loop Rotate to remove Loop Guards.
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/Transforms/Utils/LoopIterCountAssumptions.h"
+#include "llvm/Analysis/AliasAnalysis.h"
+#include "llvm/Analysis/AssumptionCache.h"
+#include "llvm/Analysis/MemorySSA.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Value.h"
+#include "llvm/Support/Casting.h"
+#include "llvm/Transforms/Utils/LoopUtils.h"
+#include "llvm/Transforms/Utils/ScalarEvolutionExpander.h"
+
+#define DEBUG_TYPE "loop-iter-count-assumptions"
+
+using namespace llvm;
+
+namespace {
+
+std::string getFunctionAndBlockNames(BasicBlock &BB) {
+  return BB.getParent()->getName().str() + " " + BB.getName().str();
+}
+
+/// Return the Branch Compare Instruction of CurrentLoop if the Loop is well
+/// formed and this pass can process the Predicate
+ICmpInst *getLoopCmpInst(Loop &CurrentLoop) {
+
+  if (CurrentLoop.isRotatedForm()) {
+    LLVM_DEBUG(dbgs() << "Loop already in rotated form. Will not add Loop "
+                         "Iteration Count assumptions.\n");
+    return nullptr;
+  }
```

Comment:
Checking: that could be an already-rotated loop with a guard? I assume we could potentially remove the guard if we add a `builtin_assume`? Nothing to do for now, I think it's good to focus on standard `for` loops :)

Reply:
No. 
The case to check against is a `do-while loop` with `iteration count metadata`.
Since this is already rotated, we don't need to add any assumptions, because our pass is redundant.

Reply:
we are so early in the optimization pipeline, we should never see an already rotated loop

Reply:
Someone could write code like
```
if (i <n) {
  do {
    foo;
    ++i;
  } while (i < n);
}
```
But yes, let's ignore that case ;)